### ISSUE TITLE
Adam implementation differs from paper (applies bias B_2 correction to \epsilon)

### ISSUE THREAD

**Issue by bmc2-stripe** [tensorflow/tensorflow#35102]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): N/A
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Ubuntu 16.04.6
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device:
- TensorFlow installed from (source or binary): binary
- TensorFlow version (use command below): v2.0.0-rc2-26-g64c3d38 2.0.0
- Python version: 3.7.3
- Bazel version (if compiling from source): N/A
- GCC/Compiler version (if compiling from source): N/A
- CUDA/cuDNN version: V10.0.130
- GPU model and memory: Tesla V100-SXM2

**Describe the current behavior**
In the Adam paper, we subtract the following quantity from our current gradient [0]:
\alpha * \hat{m_t} / (\sqrt{v_t / (1-\beta^t_2)}  + \epsilon)

<img width="475" alt="Screen Shot 2019-12-13 at 2 32 58 PM" src="https://user-images.githubusercontent.com/54961543/70836600-7f490780-1db5-11ea-9669-27c50fe48cae.png">

The Tensorflow implementation subtracts a subtly different quantity ([1]):
\alpha * \hat{m_t} * \sqrt{1-\beta^T_2} / (\sqrt{v_t} + \epsilon)

<img width="438" alt="Screen Shot 2019-12-13 at 2 34 36 PM" src="https://user-images.githubusercontent.com/54961543/70836654-c1724900-1db5-11ea-9260-5fc0678f6f39.png">

The difference between the two expressions is that in the first, we de-bias only the moving average of the squared gradient, v_t. In the second, this bias correction is also applied to \epsilon. This manifests as scaling up epilson quite a lot in very early training steps, reducing the magnitude of the gradient update.

Note that the same bug was present in PyTorch prior to v1.3. It was fixed in this PR: https://github.com/pytorch/pytorch/pull/22628. That PR description provides a useful visualization.

**Describe the expected behavior**
Implement the algorithm as described in the paper. If the old implementation is necessary to preserve back-compat, providing a flag to trigger the correct implementation would be most helpful.

[0] https://arxiv.org/pdf/1412.6980.pdf, see final 2 lines of Algorithm 1
[1] Notable lines in TF implementation regarding this issue: 
https://github.com/tensorflow/tensorflow/blob/r2.0/tensorflow/python/keras/optimizer_v2/adam.py#L162

https://github.com/tensorflow/tensorflow/blob/r2.0/tensorflow/python/keras/optimizer_v2/adam.py#L245

https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/training_ops.cc#L373

**Comment by tanzhenyu** [comment]
Thanks for the report. We have noticed this, and decided we can only make behavior change in 3.0

**Comment by alextp** [comment]
Before 3.0 we can flag-gate this behavior, though, and warn if the incompatible-with-the-paper flag is used.

@tanzhenyu can you implement the right behavior and gate it behind an optimizer constructor argument / different class?

**Comment by tanzhenyu** [comment]
> Before 3.0 we can flag-gate this behavior, though, and warn if the incompatible-with-the-paper flag is used.
> 
> @tanzhenyu can you implement the right behavior and gate it behind an optimizer constructor argument / different class?

Sounds good.

**Comment by saikumarchalla** [comment]
@bmc2-stripe Closing as duplicate.Can you please take a look at this [issue](https://github.com/pytorch/pytorch/pull/22628).Thanks!

**Comment by bmc2-stripe** [comment]
@saikumarchalla The issue you linked to is a PyTorch issue. I don't think this should have been closed.

**Comment by zhedongzheng** [comment]
This issue makes senses, and pytorch team has already fixed this.

**Comment by SamuelMarks** [comment]
I just saw this issue, was about to report it myself. Until 3.0 is available, it might be worth mentioning this issue in the docstring of `class Adam` so that those expecting paper-correct results will have a note to refer to (and won't be expected to compare the research paper with the TensorFlow implementation).

**Comment by sushreebarsa** [comment]
@bmc2-stripe It looks like you are using an older Version of Tensorflow . Many bugs have been fixed in the latest version. Could you please execute your code using Latest Version 2.5 and let us know if the issue still persists? Thanks!

**Comment by bmc2-stripe** [comment]
Hi, @sushreebarsa, I've verified that the incorrect formula is still being applied in master. Notably, the comment here references exactly the formula I listed above (which is not consistent with the paper):
https://github.com/memo/ofxMSATensorFlow/blob/master/libs/tensorflow/include/tensorflow/cc/ops/training_ops.h#L1008

**Comment by tensorflowbutler** [comment]
Hi There,

 This is a stale issue. As you are using an older version of tensorflow, we are checking to see if you still need help on this issue. Please test the issue with the latest TensorFlow (TF2.7 and tf-nightly). If the issue still persists with the newer versions of TF, please feel free to open it in [keras-team/keras](https://github.com/keras-team/keras/issues) repository by providing details about the issue and a standalone code to reproduce the issue. Thanks! 

 Please note that Keras development has moved to a separate Keras-team/keras repository to focus entirely on only Keras. Thanks!
